keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1257 @@
+as a class-based resource on the level of container orchestration. This enables
+configuring I/O scheduler priority and throttling I/O bandwidth per workload.
+Having the support for class-based resources in place, it will provide a
+framework for the future, for instance class-based network or memory type
+prioritization.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+RDT implements a class-based mechanism for controlling the cache and memory
+bandwidth QoS of applications, providing a tool for mitigating noisy neighbors
+and fulfilling SLAs. In Linux control happens via resctrl -- a
+pseudo-filesystem provided by the kernel which makes it virtually agnostic of
+the hardware architecture. The OCI runtime-spec has supported Intel RDT for a
+while already. Other hardware vendors have comparable technologies which use
+the same resctrl interface.
+
+The Linux Block IO controller parameters depend very heavily on the underlying
+hardware and system configuration (device naming/numbering, IO scheduler
+configuration etc) which makes it very impractical to control from the Pod spec
+level. In order to hide this complexity the concept of blockio classes is being
+added to the container runtimes (CRI-O and containerd). A system administrator
+is able to configure blockio controller parameters on per-class basis and the
+classes are then made available for CRI clients.
+
+Currently, there is no mechanism in Kubernetes to use these types of resources
+. CRI-O and containerd runtimes have support for RDT and blockio classes and
+they provide an bridge-gap user interface through special pod annotations. We
+would like to eventually get these types of resources first class citizen and
+properly supported in Kubernetes, providing visibility, a well-defined user
+interface, and permission controls.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Make it possible to request class resources
+  - Support RDT class assignment of containers. This is already supported by
+    the containerd and CRI-O runtime and part of the OCI runtime-spec
+  - Support blockio class assignment of containers.
+- Make the extensions flexible, enabling simple addition of other class-based
+  resource types in the future.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Interface for configuring the class-based resources.
+- Enumerating possible (class) resource types or their detailed behavior
```

Comment:
some behavior agreements are needed I think. Do we assume they are immutable? Or kubelet needs to have a way to find out that certain resource class is not available any longer and pod has to be deleted? This may affect CRI API design